### PR TITLE
Create custom styles for admin

### LINKED ISSUE #333: Custom styles for admin
**Is your feature request related to a problem? Please describe.**
Right now the custom styles only apply to the public part

**Describe the solution you'd like**
I wish you could make custom styles in the configuration board

**Describe alternatives you've considered**
n/a

**Additional context**
Paid by aFFaC

Mockups:

Screenshot from 2024-10-14 13-55-48

imatge

Specs:
- [x] Divide the current custom styles menu in two parts, same as has been done for custom fields (check screenshots)
- [x] Add a custom styles editor in the admin. Specs for this.
- [ ] Apply custom styles in the admin bye adding the `_custom_styles.html.erb` partial

### PR DESCRIPTION
**Pull request by ElviaBth**
#### :tophat: What? Why?
Create custom styles for admin. Divide the current custom styles menu in two parts, one for the public frontend and the other for the admin panel.

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to #?
- Fixes #333

### :camera: Screenshots
image
menu
image
public frontend
image
admin panel

:hearts: Thank you!

### CODE REVIEW

File: app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface [hunk 1]
```diff
@@ -3,6 +3,10 @@
 <%= render partial: "layouts/decidim/decidim_awesome/awesome_config" %>
 <% append_stylesheet_pack_tag("decidim_admin_decidim_awesome_global") %>
 <% append_javascript_pack_tag("decidim_admin_decidim_awesome_global") %>
+
+<% append_stylesheet_pack_tag "decidim_decidim_awesome", media: "all" %>
```
Comment (microstudi) [inline app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface:7]: this is the admin, I think this line does not belong here

File: spec/commands/admin/destroy_scoped_style_spec.rb [hunk 2]
```diff
@@ -50,7 +50,7 @@ module Admin
           expect { subject.call }.to broadcast(:ok)
 
           expect(AwesomeConfig.find_by(organization:, var: :scoped_styles).value.keys).not_to include("foo")
-          expect(AwesomeConfig.find_by(organization:, var: :scoped_style_foo)).to be_nil
+          expect(AwesomeConfig.find_by(organization:, var: :scoped_styles_foo)).to be_nil
```
Comment (microstudi) [inline spec/commands/admin/destroy_scoped_style_spec.rb:53]:
```suggestion
          expect(AwesomeConfig.find_by(organization:, var: :scoped_style_foo)).to be_nil
```

sub-vars are always singular

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+        within ".scoped_styles_container[data-key=\"foo\"]" do
+          accept_confirm { click_link_or_button "Remove this CSS box" }
+        end
+
+        expect(page).to have_admin_callout("removed successfully")
+        expect(page).to have_content("body {background: blue;}")
+        expect(page).not_to have_content("body {background: red;}")
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_foo)).not_to be_present
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_styles_bar)).to be_present
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:98]:
```suggestion
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar)).to be_present
```

singular

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        }
+      end
+
+      it "adds a new config helper var" do
+        click_on 'Add a new "admin panel" CSS box'
+
+        expect(page).to have_content("Processes")
+
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_styles_bar)).to be_present
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_styles_bar).constraints.first.settings).to eq("participatory_space_manifest" => "participatory_processes")
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:116]:
```suggestion
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar)).to be_present
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar).constraints.first.settings).to eq("participatory_space_manifest" => "participatory_processes")
```
Singular

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+          end
+
+          visit decidim_admin_decidim_awesome.config_path(:scoped_admin_styles)
+
+          within ".scoped_styles_container[data-key=\"bar\"] .constraints-editor" do
+            expect(page).not_to have_content("Processes")
+          end
+
+          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_styles_bar)).to be_present
+          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar).constraints).not_to be_present
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:158]:
```suggestion
          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar)).to be_present
          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_admin_style_bar).constraints).not_to be_present
```
singular

File: spec/system/admin/admin_manages_scoped_styles_spec.rb [hunk 6]
```diff
@@ -97,7 +96,7 @@
         expect(page).to have_content("body {background: blue;}")
         expect(page).not_to have_content("body {background: red;}")
         expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_foo)).not_to be_present
-        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_styles_bar)).to be_present
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_styles_spec.rb:99]:
```suggestion
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
```

File: spec/system/admin/admin_manages_scoped_styles_spec.rb [hunk 7]
```diff
@@ -110,19 +109,12 @@
...
-
-        within ".scoped_styles_container[data-key=\"foo\"] .constraints-editor" do
-          expect(page).to have_content("Processes")
-        end
+        expect(page).to have_content("Processes")
 
-        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
-        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar).constraints.first.settings).to eq("participatory_space_manifest" => "participatory_processes")
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_styles_bar)).to be_present
+        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_styles_bar).constraints.first.settings).to eq("participatory_space_manifest" => "participatory_processes")
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_styles_spec.rb:117]:
```suggestion
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
        expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar).constraints.first.settings).to eq("participatory_space_manifest" => "participatory_processes")
```

File: spec/system/admin/admin_manages_scoped_styles_spec.rb [hunk 8]
```diff
@@ -132,27 +124,38 @@
...
 
-          visit decidim_admin_decidim_awesome.config_path(:styles)
+          visit decidim_admin_decidim_awesome.config_path(:scoped_styles)
 
           within ".scoped_styles_container[data-key=\"bar\"] .constraints-editor" do
             expect(page).not_to have_content("Processes")
           end
 
-          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
+          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_styles_bar)).to be_present
```
Comment (microstudi) [inline spec/system/admin/admin_manages_scoped_styles_spec.rb:158]:
```suggestion
          expect(Decidim::DecidimAwesome::AwesomeConfig.find_by(organization:, var: :scoped_style_bar)).to be_present
```

**Review by microstudi — CHANGES_REQUESTED** [review]
Good job.

To finish it, please check the comments, there are some inconsistencies in the constrains plural names.

On the other hand, the spec `spec/system/admin/admin_manages_scoped_admin_styles_spec.rb` is repeating a lot of code, from `spec/system/admin/admin_manages_scoped_styles_spec.rb`. Can you do the same as is done in  `spec/system/admin/admin_manages_proposal_custom_fields_spec.rb` using shared examples?

And, finally, please update the screenshots and explanations from the README.md file. Probably editing the point 8 with two subsection 8.1 (public styles) and 8.2 (admin styles)

File: spec/system/admin/admin_manages_scoped_styles_spec.rb (hunk 7 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_styles_spec.rb:117]: Done

File: spec/system/admin/admin_manages_scoped_styles_spec.rb (hunk 6 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_styles_spec.rb:99]: Done

File: spec/system/admin/admin_manages_scoped_styles_spec.rb (hunk 8 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_styles_spec.rb:158]: Done

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb (hunk 5 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:158]: Done

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb (hunk 4 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:116]: Done

File: spec/system/admin/admin_manages_scoped_admin_styles_spec.rb (hunk 3 above)
Comment (ElviaBth) [reply spec/system/admin/admin_manages_scoped_admin_styles_spec.rb:98]: Done

File: app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface (hunk 1 above)
Comment (ElviaBth) [reply app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface:7]: Done

File: spec/commands/admin/destroy_scoped_style_spec.rb (hunk 2 above)
Comment (ElviaBth) [reply spec/commands/admin/destroy_scoped_style_spec.rb:53]: Done

File: app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface [hunk 9]
```diff
@@ -3,6 +3,9 @@
 <%= render partial: "layouts/decidim/decidim_awesome/awesome_config" %>
 <% append_stylesheet_pack_tag("decidim_admin_decidim_awesome_global") %>
 <% append_javascript_pack_tag("decidim_admin_decidim_awesome_global") %>
+
+<%= render(partial: "layouts/decidim/decidim_awesome/admin/custom_styles") if awesome_scoped_admin_styles.present? %>
```
Comment (microstudi) [inline app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface:7]:
```suggestion
```
Let's remove this from this file and create a new deface file:

```
<!-- insert_after "erb[loud]:contains('stylesheet_pack_tag("decidim_core", "decidim_admin")')" -->

<%= render(partial: "layouts/decidim/decidim_awesome/admin/custom_styles") if awesome_scoped_admin_styles.present? %>

```

File: app/overrides/layouts/decidim/admin/_header/add_awesome_custom_styles.html.erb.deface [hunk 10]
```diff
@@ -0,0 +1,3 @@
+<!-- insert_after "erb[loud]:contains('stylesheet_pack_tag("decidim_core", "decidim_admin")')" -->
```
Comment (microstudi) [inline app/overrides/layouts/decidim/admin/_header/add_awesome_custom_styles.html.erb.deface:1]:
```suggestion
<!-- insert_after "erb[loud]:contains('snippets.display(:head)')" -->
```

File: app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface (hunk 9 above)
Comment (ElviaBth) [reply app/overrides/layouts/decidim/admin/_header/add_awesome_tags.html.erb.deface]: Done

File: app/overrides/layouts/decidim/admin/_header/add_awesome_custom_styles.html.erb.deface (hunk 10 above)
Comment (ElviaBth) [reply app/overrides/layouts/decidim/admin/_header/add_awesome_custom_styles.html.erb.deface]: Done

**Review by microstudi — APPROVED** [review]
great work!

**Pull request merged**